_posts/2024-07-01-zio-fullstack-webapp.md

```diff
@@ -0,0 +1,844 @@
+* **(2)** An EventBus that will receive the User created
+* **(3)** A Form that will bind the Person to the UI see [here](https://cheleb.github.io/laminar-form-derivation/docs/index.html)
+* **(4)** A Button that will call the PersonEndpoint.createEndpoint with the current value of the Person Var
+* **(5)** A div that will display the current value of the Person Var (debug purpose)
+* **(6)** A div that will display the User created, reacting to the EventBus events.
+
+This is a lot of magic in a single page.
+
+
+# Interacting with JavaScript ecosystem
+
+We can leverage the JavaScript ecosystem in a type safe manner. 
+
+Either by using Scala.js facade or by using the JavaScript API directly.
+
+To illustrate this,  we are using a Laminar binding to a SAP UI5 component.
+
+But more interestingly, you will find in the project a facade to the [Chart.js](https://www.chartjs.org/) library.
+
+This being is generated by [Scalablytyped](https://scalablytyped.org/), a tool that will generate Scala.js facade from TypeScript definition.
+
+This sample is a shamy cut and paste from Sébastien Doeraene demonstration: [Getting started with Scala.js, Laminar and ScalablyTyped](https://www.youtube.com/watch?v=UePrOa_1Am8)
+
+```sbt
+addSbtPlugin("org.scalablytyped.converter" % "sbt-converter" % "1.0.0-beta44")
+```
+
+Some configuration, devDependencies point to TypeScript whereas runtime dependencies point javascript:
+```json
+{
+  "name": "scalablytyped",
+  "private": true,
+  "version": "0.0.1",
+  "type": "module",
+  "dependencies": {
+    "chart.js": "2.9.4"
+  },
+  "devDependencies": {
+    "typescript": "5.4.5",
+    "@types/chart.js": "2.9.29"
+  }
+}
+```
+
+In this example Chart.js binding are generated from the TypeScript definition and fully accessible from ScalaJS at compile time, whereas the chart.js will be used at runtime.
+Generated classes are pushed in your local (ivy) repo, hence generated only one.
+
+Notice, this package.json is not the application one, because I want to limit this binding generation to strict necessary.
+
+Then you can refer to TypeScript library very easily, see [here](https://github.com/cheleb/zio-laminar-demo/blob/master/modules/client/src/main/scala/com/example/ziolaminardemo/app/demos/ScalablytypedDemoPage.scala#L103).
+
+
+# Short conclusion
+
+This is a very simple example, but it shows how ZIO, Laminar and Tapir can be combined to build a full stack webapp. 
+
+* ZIO provides a type safe way to handle effects, and to build a full stack application both on the client and the server side.
+
+* Laminar provides a type safe way to build a reactive UI.
+
+* Tapir provides a type safe way to define API endpoints.
+
+
+This is a very powerful combination, that allows to build a full stack webapp in a type safe manner. And in the mean time we can leverage the JavaScript ecosystem, without compromising type safety, architecture and scalability of the application.
+
+
+# Roadmap
```

Comment:
Let's cut this Markdown to a sentence. There's no need to add new H1, H2 headers, "I failed" etc. You're doing great, and you want the reader to know that you'll do more. Here's an example:

"The g8 starter project needs an example of JWT token handling, splitting the JavaScript into multiple files, and maybe an example of WASM now that Scala.js has a first full implementation of WasmGC (add link). This is the subject of a future article."